src/compile/decode.js

```diff
@@ -132,14 +198,17 @@ function compileDecode (m, resolve, enc) {
 
           if (val) { // is enum
             if (field.repeated) {
-              obj[name] = []
+              def = []
             } else {
               def = (def && val[def]) ? val[def].value : val[Object.keys(val)[0]].value
-              obj[name] = parseInt(def || 0, 10)
+              def = parseInt(def || 0, 10)
             }
           } else {
-            obj[name] = defaultValue(field, def)
+            defaultVal = defaultValue(field, def)
+            def = coerceValue(field, def)
```

Comment:
```suggestion
            def = def !== undefined ? defaultValue(field, def) : undefined
```

Or something like that? We’re want to preserve undefined “values” right? Would be good not to repeat the majority of that method again!

Reply:
The problem (as I saw it) was that `defaultValue` was coercing the value if it was set or returning a field-specific zero value if it wasn't set.  I think this violates the single responsibility principal.